src/mlir/consumer.cc

```diff
@@ -0,0 +1,762 @@
+    for (auto sub : lambda->body)
+    {
+      auto node = consumeNode(sub);
+      if (auto err = node.takeError())
+        return std::move(err);
+      last = *node;
+    }
+    return last;
+  }
+
+  llvm::Expected<Value> ASTConsumer::consumeSelect(Ast ast)
+  {
+    auto select = nodeAs<Select>(ast);
+    assert(select && "Bad Node");
+    auto loc = getLocation(ast);
+
+    Value lhs, rhs;
+
+    // This is either:
+    //  * the RHS of a binary operator
+    //  * the argument of a unary operator
+    //  * the arguments of the function call as a tuple
+    if (select->args)
+    {
+      auto rhsNode = consumeNode(select->args);
+      if (auto err = rhsNode.takeError())
+        return std::move(err);
+      rhs = *rhsNode;
+    }
+
+    // FIXME: "special case" return for now, to make it work without method
+    // call. There's a bug in the current AST that doesn't create a "last" value
+    // in some cases, so we add an explicit "return" to force it.
+    if (select->typenames[0]->location.view() == "return")
+    {
+      auto thisFunc =
+        dyn_cast<FuncOp>(builder().getInsertionBlock()->getParentOp());
+      if (thisFunc.getType().getNumResults() > 0)
+        rhs = gen.AutoLoad(loc, rhs, thisFunc.getType().getResult(0));
+      return rhs;
+    }
+
+    // This is either:
+    //  * the LHS of a binary operator
+    //  * the selector for a static/dynamic call of a class member
+    if (select->expr)
+    {
+      auto lhsNode = consumeNode(select->expr);
+      if (auto err = lhsNode.takeError())
+        return std::move(err);
+      lhs = *lhsNode;
+    }
+
+    // Dynamic selector, for accessing a field or calling a method
+    if (auto structTy = gen.getPointedStructType(lhs))
+    {
+      // Loading fields, we calculate the offset to load based on the field name
+      auto [offset, elmTy, found] =
+        getField(structTy, select->typenames[0]->location.view());
+      if (found)
+      {
+        // Convert the address of the structure to the address of the element
+        return gen.GEP(loc, lhs, offset);
+      }
+
+      // FIXME: Implement dynamic dispatch of methods
+      assert(false && "Dynamic method call not implemented yet");
+    }
+
+    // Typenames indicate the context and the function name
+    llvm::SmallVector<llvm::StringRef, 3> scope;
+    size_t end = select->typenames.size() - 1;
+    for (size_t i = 0; i < end; i++)
+    {
+      scope.push_back(select->typenames[i]->location.view());
+    }
+    std::string opName =
+      mangleName(select->typenames[end]->location.view(), scope);
+
+    // Check the function table for a symbol that matches the opName
+    if (auto funcOp = gen.lookupSymbol<FuncOp>(opName))
+    {
+      llvm::SmallVector<Value, 1> args;
+      // Here it's guaranteed the lhs is not a selector (handled above), so if
+      // there is one, it's the first argument of a function call.
+      if (lhs)
+      {
+        args.push_back(lhs);
+      }
+      if (rhs)
```

Comment:
I find it quite hard to follow this block.  It would help to have a block comment here explaining what it's trying to do.